numpy/random/_generator.pyx

```diff
@@ -559,39 +558,39 @@ cdef class Generator:
             high = low
             low = 0
 
-        dt = np.dtype(dtype)
-        key = dt.name
-        if key not in _integers_types:
-            raise TypeError('Unsupported dtype "%s" for integers' % key)
-        if not dt.isnative:
-            raise ValueError('Providing a dtype with a non-native byteorder '
-                             'is not supported. If you require '
-                             'platform-independent byteorder, call byteswap '
-                             'when required.')
+        _dtype = np.dtype(dtype)
 
         # Implementation detail: the old API used a masked method to generate
         # bounded uniform integers. Lemire's method is preferable since it is
         # faster. randomgen allows a choice, we will always use the faster one.
         cdef bint _masked = False
 
-        if key == 'int32':
+        if _dtype == np.int32:
             ret = _rand_int32(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
-        elif key == 'int64':
+        elif _dtype == np.int64:
             ret = _rand_int64(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
-        elif key == 'int16':
+        elif _dtype == np.int16:
             ret = _rand_int16(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
-        elif key == 'int8':
+        elif _dtype == np.int8:
             ret = _rand_int8(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
-        elif key == 'uint64':
+        elif _dtype == np.uint64:
             ret = _rand_uint64(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
-        elif key == 'uint32':
+        elif _dtype == np.uint32:
             ret = _rand_uint32(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
-        elif key == 'uint16':
+        elif _dtype == np.uint16:
             ret = _rand_uint16(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
-        elif key == 'uint8':
+        elif _dtype == np.uint8:
             ret = _rand_uint8(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
-        elif key == 'bool':
+        elif _dtype == np.bool_:
             ret = _rand_bool(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
```

Comment:
A lookup table might be faster here, as:
```
_rand_int_funcs = {
    np.dtype(np.bool_): _rand_bool,
    ...
```
then
```
try:
    func = _rand_int_funcs[dt]
except KeyError:
    raise ValueError(...) from None
else:
    return func(...)

Reply:
I tried your suggestion, but I have an error during compilation. I don't have much experience with Cython, could you take a look? 
https://gist.github.com/przemb/95d6782dada6f1cfb05ff53c810577d4


btw: I thought about [overlading](https://www.python.org/dev/peps/pep-3124/#overloading-generic-functions), but I guess that it is not possible? (in order to keep backward compatibility)

Reply:
`_rand_bool` etc. are C functions (essentially) and cannot be used as dict values.

Reply:
The `if/elif` chain should be fine.

Reply:
The advantage of the dict approach is that the `np.dtype` objects can be pre-computed, rather than `dt == np.float32` being expanded to `dt == np.dtype(np.float32)` on each call.

@rkern's explanation of why this doesn't work is correct, and it looks like it's more trouble than it's worth to make something like that work - so I agree that what you have is fine.